File: tests/security_authorization/test_security_authorization.cpp
#include <stdio.h>
#include <string.h>

#include "security_authorization_core.h"

namespace sec = myactuator::security;

namespace {

int failures = 0;
#define CHECK(value)                                                         \
    do {                                                                     \
        if (!(value)) {                                                      \
            fprintf(stderr, "CHECK failed at %s:%d: %s\n", __FILE__,       \
                    __LINE__, #value);                                       \
            ++failures;                                                      \
        }                                                                    \
    } while (0)

sec::Digest Digest(uint8_t value) {
    sec::Digest output = {};
    for (size_t index = 0U; index < sec::kDigestSize; ++index)
        output.bytes[index] = value;
    return output;
}

sec::AuthorizationPolicy Policy() {
    sec::AuthorizationPolicy value = {};
    value.expected_config_digest = Digest(4U);
    value.expected_source_generation_digest = Digest(5U);
    value.expected_graph_generation_digest = Digest(6U);
    value.physical_actuation_enabled = false;
    value.remote_physical_actuation_enabled = false;
    value.remote_administration_enabled = false;
    return value;
}

sec::IdentityAssertion Identity() {
    sec::IdentityAssertion value = {};
    value.actor_digest = Digest(1U);
    value.session_digest = Digest(2U);
    value.authentication_context_digest = Digest(3U);
    value.role = sec::Role::OPERATOR;
    value.authenticated = true;
    value.revoked = false;
    value.valid_from_ns = 100U;
    value.valid_until_ns = 1000U;
    return value;
}

sec::ApprovalAssertion Approval() {
    sec::ApprovalAssertion value = {};
    value.actor_digest = Digest(9U);
    value.authentication_context_digest = Digest(10U);
    value.role = sec::Role::EVIDENCE_REVIEWER;
    value.authenticated = true;
    value.revoked = false;
    value.valid_from_ns = 100U;
    value.valid_until_ns = 1000U;
    value.scope_digest = Digest(8U);
    return value;
}

sec::AuthorizationRequest Request() {
    sec::AuthorizationRequest value = {};
    value.action = sec::Action::SUBMIT_MOTION;
    value.target = sec::Target::PHYSICAL_REMOTE;
    value.safety_state = sec::SafetyState::ENABLED;
    value.session_digest = Digest(2U);
    value.correlation_digest = Digest(7U);
    value.sequence = 1U;
    value.now_ns = 500U;
    value.config_digest = Digest(4U);
    value.source_generation_digest = Digest(5U);
    value.graph_generation_digest = Digest(6U);
    value.artifact_digest = Digest(8U);
    value.lease_valid = true;
    value.safety_admission_ready = true;
    value.local_presence_verified = true;
    value.artifact_integrity_verified = true;
    value.rollback_guard_verified = true;
    value.approval_present = true;
    value.approval = Approval();
    return value;
}

sec::DecisionCode RunVariant(const char* variant) {
    sec::AuthorizationPolicy policy = Policy();
    sec::IdentityAssertion identity = Identity();
    sec::AuthorizationRequest request = Request();
    size_t normal_capacity = 16U;
    bool prefill = false;
    bool replay = false;
    bool normal_ready = true;
    bool safe_ready = true;

    if (strcmp(variant, "observer_read") == 0) {
        identity.role = sec::Role::OBSERVER;
        request.action = sec::Action::READ_STATE;
        request.target = sec::Target::OFFLINE;
    } else if (strcmp(variant, "diagnostic_read") == 0) {
        identity.role = sec::Role::DIAGNOSTIC_OPERATOR;
        request.action = sec::Action::READ_DIAGNOSTICS;
    } else if (strcmp(variant, "simulation_motion") == 0) {
        request.target = sec::Target::SIMULATION;
    } else if (strcmp(variant, "physical_remote_enabled") == 0) {
        policy.physical_actuation_enabled = true;
        policy.remote_physical_actuation_enabled = true;
    } else if (strcmp(variant, "safe_disable") == 0) {
        request.action = sec::Action::REQUEST_DISABLE;
    } else if (strcmp(variant, "config_activation") == 0) {
        identity.role = sec::Role::CONFIGURATION_MANAGER;
        request.action = sec::Action::ACTIVATE_CONFIG;
        request.target = sec::Target::PHYSICAL_LOCAL;
    } else if (strcmp(variant, "firmware_activation") == 0) {
        identity.role = sec::Role::FIRMWARE_MANAGER;
        request.action = sec::Action::ACTIVATE_FIRMWARE;
        request.target = sec::Target::PHYSICAL_LOCAL;
    } else if (strcmp(variant, "evidence_submit") == 0) {
        identity.role = sec::Role::EVIDENCE_REVIEWER;
        request.action = sec::Action::SUBMIT_EVIDENCE;
        request.target = sec::Target::OFFLINE;
    } else if (strcmp(variant, "unauthenticated") == 0) {
        identity.authenticated = false;
    } else if (strcmp(variant, "revoked") == 0) {
        identity.revoked = true;
    } else if (strcmp(variant, "session_not_yet_valid") == 0) {
        request.now_ns = 99U;
    } else if (strcmp(variant, "session_expired") == 0) {
        request.now_ns = 1000U;
    } else if (strcmp(variant, "session_mismatch") == 0) {
        request.session_digest = Digest(11U);
    } else if (strcmp(variant, "replay") == 0) {
        request.target = sec::Target::SIMULATION;
        replay = true;
    } else if (strcmp(variant, "role_denied") == 0) {
        identity.role = sec::Role::DIAGNOSTIC_OPERATOR;
        request.target = sec::Target::SIMULATION;
    } else if (strcmp(variant, "physical_disabled") == 0) {
    } else if (strcmp(variant, "remote_physical_disabled") == 0) {
        policy.physical_actuation_enabled = true;
    } else if (strcmp(variant, "remote_admin_disabled") == 0) {
        identity.role = sec::Role::CONFIGURATION_MANAGER;
        request.action = sec::Action::STAGE_CONFIG;
    } else if (strcmp(variant, "local_presence") == 0) {
        identity.role = sec::Role::CONFIGURATION_MANAGER;
        request.action = sec::Action::ACTIVATE_CONFIG;
        request.target = sec::Target::PHYSICAL_LOCAL;
        request.local_presence_verified = false;
    } else if (strcmp(variant, "config_missing") == 0) {
        request.target = sec::Target::SIMULATION;
        request.config_digest = Digest(0U);
    } else if (strcmp(variant, "config_mismatch") == 0) {
        request.target = sec::Target::SIMULATION;
        request.config_digest = Digest(11U);
    } else if (strcmp(variant, "source_missing") == 0) {
        request.target = sec::Target::SIMULATION;
        request.source_generation_digest = Digest(0U);
    } else if (strcmp(variant, "source_mismatch") == 0) {
        request.target = sec::Target::SIMULATION;
        request.source_generation_digest = Digest(11U);
    } else if (strcmp(variant, "graph_missing") == 0) {
        request.target = sec::Target::SIMULATION;
        request.graph_generation_digest = Digest(0U);
    } else if (strcmp(variant, "graph_mismatch") == 0) {
        request.target = sec::Target::SIMULATION;
        request.graph_generation_digest = Digest(11U);
    } else if (strcmp(variant, "lease_missing") == 0) {
        request.target = sec::Target::SIMULATION;
        request.lease_valid = false;
    } else if (strcmp(variant, "safety_missing") == 0) {
        request.target = sec::Target::SIMULATION;
        request.safety_admission_ready = false;
    } else if (strcmp(variant, "artifact_integrity") == 0) {
        identity.role = sec::Role::CONFIGURATION_MANAGER;
        request.action = sec::Action::STAGE_CONFIG;
        request.target = sec::Target::OFFLINE;
        request.artifact_integrity_verified = false;
    } else if (strcmp(variant, "rollback_missing") == 0) {
        identity.role = sec::Role::CONFIGURATION_MANAGER;
        request.action = sec::Action::ACTIVATE_CONFIG;
        request.target = sec::Target::OFFLINE;
        request.rollback_guard_verified = false;
    } else if (strcmp(variant, "approval_missing") == 0) {
        identity.role = sec::Role::CONFIGURATION_MANAGER;
        request.action = sec::Action::ACTIVATE_CONFIG;
        request.target = sec::Target::OFFLINE;
        request.approval_present = false;
    } else if (strcmp(variant, "approval_same_actor") == 0) {
        identity.role = sec::Role::CONFIGURATION_MANAGER;
        request.action = sec::Action::ACTIVATE_CONFIG;
        request.target = sec::Target::OFFLINE;
        request.approval.actor_digest = Digest(1U);
    } else if (strcmp(variant, "approval_wrong_role") == 0) {
        identity.role = sec::Role::CONFIGURATION_MANAGER;
        request.action = sec::Action::ACTIVATE_CONFIG;
        request.target = sec::Target::OFFLINE;
        request.approval.role = sec::Role::SAFETY_OPERATOR;
    } else if (strcmp(variant, "approval_scope") == 0) {
        identity.role = sec::Role::CONFIGURATION_MANAGER;
        request.action = sec::Action::ACTIVATE_CONFIG;
        request.target = sec::Target::OFFLINE;
        request.approval.scope_digest = Digest(11U);
    } else if (strcmp(variant, "audit_unavailable") == 0) {
        request.target = sec::Target::SIMULATION;
        normal_ready = false;
    } else if (strcmp(variant, "safe_audit_unavailable") == 0) {
        request.action = sec::Action::REQUEST_DISABLE;
        safe_ready = false;
    } else if (strcmp(variant, "audit_full") == 0) {
        identity.role = sec::Role::DIAGNOSTIC_OPERATOR;
        request.action = sec::Action::READ_STATE;
        request.target = sec::Target::OFFLINE;
        request.sequence = 2U;
        normal_capacity = 1U;
        prefill = true;
    } else if (strcmp(variant, "invalid_request") == 0) {
        identity.authenticated = false;
        request.correlation_digest = Digest(0U);
    } else {
        CHECK(false);
    }

    sec::AuthorizationEngine engine(policy, identity, normal_capacity, 2U);
    engine.setAuditHealth(normal_ready, safe_ready);
    if (prefill) {
        sec::AuthorizationRequest first = Request();
        first.action = sec::Action::READ_STATE;
        first.target = sec::Target::OFFLINE;
        CHECK(engine.authorize(first).code ==
              sec::DecisionCode::PASS_TO_NEXT_GATE);
    }
    if (replay)
        CHECK(engine.authorize(request).code ==
              sec::DecisionCode::PASS_TO_NEXT_GATE);
    const sec::AuthorizationResult result = engine.authorize(request);
    CHECK(!result.motion_authorized);
    CHECK(result.proceed_to_next_gate ==
          (result.code == sec::DecisionCode::PASS_TO_NEXT_GATE));
    return result.code;
}

bool Extract(const char* line, const char* prefix, char* output,
             size_t output_size) {
    const char* start = strstr(line, prefix);
    if (start == NULL) return false;
    start += strlen(prefix);
    const char* end = strchr(start, '"');
    if (end == NULL) return false;
    const size_t length = static_cast<size_t>(end - start);
    if (length == 0U || length >= output_size) return false;
    memcpy(output, start, length);
    output[length] = '\0';
    return true;
}

void TestSharedCorpus() {
    FILE* file =
        fopen("tests/security_authorization/golden_authorization.jsonl", "rb");
    CHECK(file != NULL);
    if (file == NULL) return;
    char line[512] = {};
    char variant[96] = {};
    char expected[96] = {};
    size_t count = 0U;
    while (fgets(line, sizeof(line), file) != NULL) {
        CHECK(strchr(line, '\n') != NULL);
        CHECK(Extract(line, "\"variant\":\"", variant, sizeof(variant)));
        CHECK(Extract(line, "\"expected_code\":\"", expected,
                      sizeof(expected)));
        const sec::DecisionCode actual = RunVariant(variant);
        CHECK(strcmp(sec::DecisionCodeName(actual), expected) == 0);
        ++count;
    }
    CHECK(!ferror(file));
    fclose(file);
    CHECK(count == 37U);
}

void TestRoleMatrixClosed() {
    static const bool expected[7][10] = {
        {true, false, false, false, false, false, false, false, false, false},
        {true, true, false, false, false, false, false, false, false, false},
        {true, true, true, true, false, false, false, false, false, false},
        {true, true, false, true, true, false, false, false, false, false},
        {true, true, false, false, false, true, true, false, false, false},
        {true, true, false, false, false, false, false, true, true, false},
        {true, true, false, false, false, false, false, false, false, true},
    };
    for (uint8_t role_index = 0U; role_index < 7U; ++role_index) {
        for (uint8_t action_index = 0U; action_index < 10U; ++action_index) {
            sec::IdentityAssertion identity = Identity();
            identity.role = static_cast<sec::Role>(role_index + 1U);
            sec::AuthorizationRequest request = Request();
            request.action = static_cast<sec::Action>(action_index + 1U);
            request.target = request.action == sec::Action::SUBMIT_MOTION
                                 ? sec::Target::SIMULATION
                                 : sec::Target::OFFLINE;
            sec::AuthorizationEngine engine(Policy(), identity, 2U, 1U);
            const sec::DecisionCode code = engine.authorize(request).code;
            CHECK((code != sec::DecisionCode::ROLE_ACTION_DENIED) ==
                  expected[role_index][action_index]);
        }
    }
}

void TestSequenceAndAuditLanes() {
    sec::IdentityAssertion diagnostic = Identity();
    diagnostic.role = sec::Role::DIAGNOSTIC_OPERATOR;
    sec::AuthorizationEngine replay_engine(Policy(), diagnostic, 3U, 1U);
    sec::AuthorizationRequest denied = Request();
    denied.target = sec::Target::SIMULATION;
    const sec::AuthorizationResult first = replay_engine.authorize(denied);
    CHECK(first.code == sec::DecisionCode::ROLE_ACTION_DENIED);
    CHECK(first.sequence_committed);
    denied.action = sec::Action::READ_STATE;
    denied.target = sec::Target::OFFLINE;
    const sec::AuthorizationResult replayed = replay_engine.authorize(denied);
    CHECK(replayed.code == sec::DecisionCode::REPLAY_OR_REORDER);
    CHECK(!replayed.sequence_committed);

    sec::AuthorizationEngine engine(Policy(), Identity(), 1U, 2U);
    sec::AuthorizationRequest request = Request();
    request.action = sec::Action::READ_DIAGNOSTICS;
    request.target = sec::Target::OFFLINE;
    CHECK(engine.authorize(request).code ==
          sec::DecisionCode::PASS_TO_NEXT_GATE);
    request.sequence = 2U;
    CHECK(engine.authorize(request).code ==
          sec::DecisionCode::AUDIT_CAPACITY_EXHAUSTED);
    for (uint64_t sequence = 3U; sequence <= 5U; ++sequence) {
        request.action = sec::Action::REQUEST_DISABLE;
        request.target = sec::Target::PHYSICAL_REMOTE;
        request.sequence = sequence;
        const sec::AuthorizationResult result = engine.authorize(request);
        CHECK(result.code == sec::DecisionCode::PASS_TO_NEXT_GATE);
        CHECK(result.audit_stored);
    }
    CHECK(engine.normalAuditSize() == 1U);
    CHECK(engine.safeAuditSize() == 2U);
    CHECK(engine.safeAuditOverwriteCount() == 1U);
    CHECK(engine.safeAuditAt(0U)->sequence == 4U);
    CHECK(engine.safeAuditAt(1U)->sequence == 5U);
}

void TestAuditContextAndClosedEnums() {
    sec::IdentityAssertion identity = Identity();
    identity.role = sec::Role::CONFIGURATION_MANAGER;
    sec::AuthorizationRequest request = Request();
    request.action = sec::Action::ACTIVATE_CONFIG;
    request.target = sec::Target::PHYSICAL_LOCAL;
    sec::AuthorizationEngine engine(Policy(), identity, 2U, 1U);
    const sec::AuthorizationResult result = engine.authorize(request);
    CHECK(result.code == sec::DecisionCode::PASS_TO_NEXT_GATE);
    CHECK(result.audit_stored);
    CHECK(sec::DigestEqual(result.audit_event.actor_digest,
                           identity.actor_digest));
    CHECK(sec::DigestEqual(result.audit_event.session_digest,
                           identity.session_digest));
    CHECK(sec::DigestEqual(result.audit_event.config_digest,
                           request.config_digest));
    CHECK(sec::DigestEqual(result.audit_event.source_generation_digest,
                           request.source_generation_digest));
    CHECK(sec::DigestEqual(result.audit_event.graph_generation_digest,
                           request.graph_generation_digest));
    CHECK(sec::DigestEqual(result.audit_event.artifact_digest,
                           request.artifact_digest));
    CHECK(result.audit_event.safety_state == sec::SafetyState::ENABLED);
    CHECK(result.audit_event.decision ==
          sec::DecisionCode::PASS_TO_NEXT_GATE);
    for (uint8_t raw = 0U;
         raw <=
         static_cast<uint8_t>(sec::DecisionCode::AUDIT_CAPACITY_EXHAUSTED);
         ++raw)
        CHECK(strcmp(sec::DecisionCodeName(
                         static_cast<sec::DecisionCode>(raw)),
                     "UNKNOWN_DECISION_CODE") != 0);
    CHECK(strcmp(sec::DecisionCodeName(
                     static_cast<sec::DecisionCode>(255U)),
                 "UNKNOWN_DECISION_CODE") == 0);
    sec::AuthorizationEngine invalid_capacity(Policy(), Identity(), 17U, 1U);
    CHECK(invalid_capacity.authorize(Request()).code ==
          sec::DecisionCode::INVALID_REQUEST);
    CHECK(sizeof(sec::AuthorizationEngine) <= 8192U);
}

}  // namespace

int main() {
    TestSharedCorpus();
    TestRoleMatrixClosed();
    TestSequenceAndAuditLanes();
    TestAuditContextAndClosedEnums();
    if (failures != 0) return 1;
    printf("SECURITY_AUTHORIZATION_CORE_OK\n");
    return 0;
}

File: tests/security_authorization/test_security_authorization.py
from __future__ import annotations

import dataclasses
import json
import unittest
from dataclasses import replace
from pathlib import Path

from myactuator_lib.security_authorization import (
    Action,
    ApprovalAssertion,
    AuthorizationEngine,
    AuthorizationPolicy,
    AuthorizationRequest,
    DecisionCode,
    IdentityAssertion,
    Role,
    SafetyState,
    Target,
)


FIXTURE = Path(__file__).with_name("golden_authorization.jsonl")
ZERO = "0" * 64
ACTOR = "1" * 64
SESSION = "2" * 64
AUTH_CONTEXT = "3" * 64
CONFIG = "4" * 64
SOURCE = "5" * 64
GRAPH = "6" * 64
CORRELATION = "7" * 64
ARTIFACT = "8" * 64
REVIEWER = "9" * 64


def policy(**changes) -> AuthorizationPolicy:
    return replace(
        AuthorizationPolicy(CONFIG, SOURCE, GRAPH),
        **changes,
    )


def identity(**changes) -> IdentityAssertion:
    return replace(
        IdentityAssertion(
            ACTOR,
            SESSION,
            AUTH_CONTEXT,
            Role.OPERATOR,
            True,
            False,
            100,
            1000,
        ),
        **changes,
    )


def approval(**changes) -> ApprovalAssertion:
    return replace(
        ApprovalAssertion(
            REVIEWER,
            "a" * 64,
            Role.EVIDENCE_REVIEWER,
            True,
            False,
            100,
            1000,
            ARTIFACT,
        ),
        **changes,
    )


def request(**changes) -> AuthorizationRequest:
    return replace(
        AuthorizationRequest(
            Action.SUBMIT_MOTION,
            Target.PHYSICAL_REMOTE,
            SafetyState.ENABLED,
            SESSION,
            CORRELATION,
            1,
            500,
            CONFIG,
            SOURCE,
            GRAPH,
            ARTIFACT,
            True,
            True,
            True,
            True,
            True,
            approval(),
        ),
        **changes,
    )


def corpus_case(variant: str):
    selected_policy = policy()
    selected_identity = identity()
    selected_request = request()
    normal_capacity = 16
    prefill = False
    normal_ready = True
    safe_ready = True
    replay = False

    if variant == "observer_read":
        selected_identity = identity(role=Role.OBSERVER)
        selected_request = request(action=Action.READ_STATE, target=Target.OFFLINE)
    elif variant == "diagnostic_read":
        selected_identity = identity(role=Role.DIAGNOSTIC_OPERATOR)
        selected_request = request(
            action=Action.READ_DIAGNOSTICS, target=Target.PHYSICAL_REMOTE
        )
    elif variant == "simulation_motion":
        selected_request = request(target=Target.SIMULATION)
    elif variant == "physical_remote_enabled":
        selected_policy = policy(
            physical_actuation_enabled=True,
            remote_physical_actuation_enabled=True,
        )
    elif variant == "safe_disable":
        selected_request = request(action=Action.REQUEST_DISABLE)
    elif variant == "config_activation":
        selected_identity = identity(role=Role.CONFIGURATION_MANAGER)
        selected_request = request(
            action=Action.ACTIVATE_CONFIG,
            target=Target.PHYSICAL_LOCAL,
        )
    elif variant == "firmware_activation":
        selected_identity = identity(role=Role.FIRMWARE_MANAGER)
        selected_request = request(
            action=Action.ACTIVATE_FIRMWARE,
            target=Target.PHYSICAL_LOCAL,
        )
    elif variant == "evidence_submit":
        selected_identity = identity(role=Role.EVIDENCE_REVIEWER)
        selected_request = request(
            action=Action.SUBMIT_EVIDENCE, target=Target.OFFLINE
        )
    elif variant == "unauthenticated":
        selected_identity = identity(authenticated=False)
    elif variant == "revoked":
        selected_identity = identity(revoked=True)
    elif variant == "session_not_yet_valid":
        selected_request = request(now_ns=99)
    elif variant == "session_expired":
        selected_request = request(now_ns=1000)
    elif variant == "session_mismatch":
        selected_request = request(session_digest="b" * 64)
    elif variant == "replay":
        selected_request = request(target=Target.SIMULATION)
        replay = True
    elif variant == "role_denied":
        selected_identity = identity(role=Role.DIAGNOSTIC_OPERATOR)
        selected_request = request(target=Target.SIMULATION)
    elif variant == "physical_disabled":
        pass
    elif variant == "remote_physical_disabled":
        selected_policy = policy(physical_actuation_enabled=True)
    elif variant == "remote_admin_disabled":
        selected_identity = identity(role=Role.CONFIGURATION_MANAGER)
        selected_request = request(action=Action.STAGE_CONFIG)
    elif variant == "local_presence":
        selected_identity = identity(role=Role.CONFIGURATION_MANAGER)
        selected_request = request(
            action=Action.ACTIVATE_CONFIG,
            target=Target.PHYSICAL_LOCAL,
            local_presence_verified=False,
        )
    elif variant == "config_missing":
        selected_request = request(target=Target.SIMULATION, config_digest=ZERO)
    elif variant == "config_mismatch":
        selected_request = request(
            target=Target.SIMULATION, config_digest="b" * 64
        )
    elif variant == "source_missing":
        selected_request = request(
            target=Target.SIMULATION, source_generation_digest=ZERO
        )
    elif variant == "source_mismatch":
        selected_request = request(
            target=Target.SIMULATION, source_generation_digest="b" * 64
        )
    elif variant == "graph_missing":
        selected_request = request(
            target=Target.SIMULATION, graph_generation_digest=ZERO
        )
    elif variant == "graph_mismatch":
        selected_request = request(
            target=Target.SIMULATION, graph_generation_digest="b" * 64
        )
    elif variant == "lease_missing":
        selected_request = request(target=Target.SIMULATION, lease_valid=False)
    elif variant == "safety_missing":
        selected_request = request(
            target=Target.SIMULATION, safety_admission_ready=False
        )
    elif variant == "artifact_integrity":
        selected_identity = identity(role=Role.CONFIGURATION_MANAGER)
        selected_request = request(
            action=Action.STAGE_CONFIG,
            target=Target.OFFLINE,
            artifact_integrity_verified=False,
        )
    elif variant == "rollback_missing":
        selected_identity = identity(role=Role.CONFIGURATION_MANAGER)
        selected_request = request(
            action=Action.ACTIVATE_CONFIG,
            target=Target.OFFLINE,
            rollback_guard_verified=False,
        )
    elif variant == "approval_missing":
        selected_identity = identity(role=Role.CONFIGURATION_MANAGER)
        selected_request = request(
            action=Action.ACTIVATE_CONFIG,
            target=Target.OFFLINE,
            approval=None,
        )
    elif variant == "approval_same_actor":
        selected_identity = identity(role=Role.CONFIGURATION_MANAGER)
        selected_request = request(
            action=Action.ACTIVATE_CONFIG,
            target=Target.OFFLINE,
            approval=approval(actor_digest=ACTOR),
        )
    elif variant == "approval_wrong_role":
        selected_identity = identity(role=Role.CONFIGURATION_MANAGER)
        selected_request = request(
            action=Action.ACTIVATE_CONFIG,
            target=Target.OFFLINE,
            approval=approval(role=Role.SAFETY_OPERATOR),
        )
    elif variant == "approval_scope":
        selected_identity = identity(role=Role.CONFIGURATION_MANAGER)
        selected_request = request(
            action=Action.ACTIVATE_CONFIG,
            target=Target.OFFLINE,
            approval=approval(scope_digest="b" * 64),
        )
    elif variant == "audit_unavailable":
        selected_request = request(target=Target.SIMULATION)
        normal_ready = False
    elif variant == "safe_audit_unavailable":
        selected_request = request(action=Action.REQUEST_DISABLE)
        safe_ready = False
    elif variant == "audit_full":
        selected_identity = identity(role=Role.DIAGNOSTIC_OPERATOR)
        selected_request = request(
            action=Action.READ_STATE,
            target=Target.OFFLINE,
            sequence=2,
        )
        normal_capacity = 1
        prefill = True
    elif variant == "invalid_request":
        selected_identity = identity(authenticated=False)
        selected_request = request(correlation_digest=ZERO)
    else:
        raise AssertionError(f"unknown corpus variant {variant}")

    engine = AuthorizationEngine(
        selected_policy,
        selected_identity,
        normal_audit_capacity=normal_capacity,
        safe_audit_capacity=2,
    )
    engine.set_audit_health(normal_ready=normal_ready, safe_ready=safe_ready)
    if prefill:
        first = engine.authorize(
            request(
                action=Action.READ_STATE,
                target=Target.OFFLINE,
                sequence=1,
            )
        )
        assert first.code == DecisionCode.PASS_TO_NEXT_GATE
    if replay:
        first = engine.authorize(selected_request)
        assert first.code == DecisionCode.PASS_TO_NEXT_GATE
    return engine.authorize(selected_request), engine


class SecurityAuthorizationTests(unittest.TestCase):
    def test_shared_corpus_is_canonical_unique_and_exact(self) -> None:
        rows = FIXTURE.read_text(encoding="utf-8").splitlines()
        self.assertEqual(37, len(rows))
        seen = set()
        for line in rows:
            row = json.loads(line)
            self.assertEqual(
                line,
                json.dumps(row, sort_keys=True, separators=(",", ":")),
            )
            self.assertNotIn(row["case"], seen)
            seen.add(row["case"])
            result, _ = corpus_case(row["variant"])
            self.assertEqual(
                row["expected_code"],
                result.code.name,
                row["case"],
            )
            self.assertFalse(result.motion_authorized)
            self.assertEqual(
                result.proceed_to_next_gate,
                result.code == DecisionCode.PASS_TO_NEXT_GATE,
            )

    def test_closed_role_matrix_has_no_wildcard_or_admin_role(self) -> None:
        expected = {
            Role.OBSERVER: {Action.READ_STATE},
            Role.DIAGNOSTIC_OPERATOR: {
                Action.READ_STATE,
                Action.READ_DIAGNOSTICS,
            },
            Role.OPERATOR: {
                Action.READ_STATE,
                Action.READ_DIAGNOSTICS,
                Action.SUBMIT_MOTION,
                Action.REQUEST_DISABLE,
            },
            Role.SAFETY_OPERATOR: {
                Action.READ_STATE,
                Action.READ_DIAGNOSTICS,
                Action.REQUEST_DISABLE,
                Action.RESET_FAULT,
            },
            Role.CONFIGURATION_MANAGER: {
                Action.READ_STATE,
                Action.READ_DIAGNOSTICS,
                Action.STAGE_CONFIG,
                Action.ACTIVATE_CONFIG,
            },
            Role.FIRMWARE_MANAGER: {
                Action.READ_STATE,
                Action.READ_DIAGNOSTICS,
                Action.STAGE_FIRMWARE,
                Action.ACTIVATE_FIRMWARE,
            },
            Role.EVIDENCE_REVIEWER: {
                Action.READ_STATE,
                Action.READ_DIAGNOSTICS,
                Action.SUBMIT_EVIDENCE,
            },
        }
        for role in Role:
            for index, action in enumerate(Action, start=1):
                selected_request = request(
                    action=action,
                    target=Target.OFFLINE,
                    sequence=index,
                )
                if action == Action.SUBMIT_MOTION:
                    selected_request = replace(
                        selected_request, target=Target.SIMULATION
                    )
                engine = AuthorizationEngine(policy(), identity(role=role))
                result = engine.authorize(selected_request)
                if action in expected[role]:
                    self.assertNotEqual(
                        DecisionCode.ROLE_ACTION_DENIED,
                        result.code,
                        (role, action),
                    )
                else:
                    self.assertEqual(
                        DecisionCode.ROLE_ACTION_DENIED,
                        result.code,
                        (role, action),
                    )
        self.assertNotIn("ADMIN", {role.name for role in Role})

    def test_denied_fresh_request_consumes_sequence_and_cannot_be_replayed(self) -> None:
        engine = AuthorizationEngine(
            policy(), identity(role=Role.DIAGNOSTIC_OPERATOR)
        )
        denied = engine.authorize(request(target=Target.SIMULATION))
        self.assertEqual(DecisionCode.ROLE_ACTION_DENIED, denied.code)
        self.assertTrue(denied.sequence_committed)
        replayed = engine.authorize(
            request(action=Action.READ_STATE, target=Target.OFFLINE)
        )
        self.assertEqual(DecisionCode.REPLAY_OR_REORDER, replayed.code)
        self.assertFalse(replayed.sequence_committed)

    def test_diagnostics_cannot_exhaust_safe_disable_audit_lane(self) -> None:
        engine = AuthorizationEngine(
            policy(),
            identity(role=Role.OPERATOR),
            normal_audit_capacity=1,
            safe_audit_capacity=2,
        )
        self.assertEqual(
            DecisionCode.PASS_TO_NEXT_GATE,
            engine.authorize(
                request(
                    action=Action.READ_DIAGNOSTICS,
                    target=Target.OFFLINE,
                    sequence=1,
                )
            ).code,
        )
        self.assertEqual(
            DecisionCode.AUDIT_CAPACITY_EXHAUSTED,
            engine.authorize(
                request(
                    action=Action.READ_DIAGNOSTICS,
                    target=Target.OFFLINE,
                    sequence=2,
                )
            ).code,
        )
        for sequence in (3, 4, 5):
            result = engine.authorize(
                request(action=Action.REQUEST_DISABLE, sequence=sequence)
            )
            self.assertEqual(DecisionCode.PASS_TO_NEXT_GATE, result.code)
            self.assertTrue(result.audit_event)
        self.assertEqual(1, len(engine.normal_audit))
        self.assertEqual(2, len(engine.safe_audit))
        self.assertEqual(1, engine.safe_audit_overwrite_count)
        self.assertEqual([4, 5], [event.sequence for event in engine.safe_audit])

    def test_audit_is_context_complete_digest_only_and_secret_free(self) -> None:
        result, engine = corpus_case("config_activation")
        self.assertEqual(DecisionCode.PASS_TO_NEXT_GATE, result.code)
        event = result.audit_event
        self.assertIsNotNone(event)
        assert event is not None
        required = {
            "actor_digest",
            "session_digest",
            "correlation_digest",
            "authentication_context_digest",
            "action",
            "target",
            "role",
            "decision",
            "safety_state",
            "sequence",
            "monotonic_time_ns",
            "config_digest",
            "source_generation_digest",
            "graph_generation_digest",
            "artifact_digest",
            "lease_valid",
            "safety_admission_ready",
        }
        fields = {field.name for field in dataclasses.fields(event)}
        self.assertTrue(required <= fields)
        forbidden_names = ("secret", "password", "credential", "private", "token")
        self.assertFalse(
            any(fragment in name for name in fields for fragment in forbidden_names)
        )
        encoded = json.dumps(dataclasses.asdict(event), sort_keys=True)
        self.assertNotIn("SEEDED-SECRET-MUST-NOT-APPEAR", encoded)
        for field in (
            event.actor_digest,
            event.session_digest,
            event.correlation_digest,
            event.authentication_context_digest,
            event.config_digest,
            event.source_generation_digest,
            event.graph_generation_digest,
            event.artifact_digest,
        ):
            self.assertRegex(field, r"^[0-9a-f]{64}$")
        self.assertEqual(1, len(engine.normal_audit))

    def test_default_policy_denies_physical_motion_but_not_offline_work(self) -> None:
        physical = AuthorizationEngine(policy(), identity()).authorize(request())
        offline = AuthorizationEngine(policy(), identity()).authorize(
            request(target=Target.SIMULATION)
        )
        self.assertEqual(
            DecisionCode.PHYSICAL_ACTUATION_DISABLED, physical.code
        )
        self.assertEqual(DecisionCode.PASS_TO_NEXT_GATE, offline.code)
        self.assertFalse(offline.motion_authorized)

    def test_audit_capacity_configuration_is_bounded(self) -> None:
        for normal_capacity, safe_capacity in ((0, 1), (17, 1), (1, 0), (1, 5)):
            with self.subTest(
                normal_capacity=normal_capacity,
                safe_capacity=safe_capacity,
            ), self.assertRaises(ValueError):
                AuthorizationEngine(
                    policy(),
                    identity(),
                    normal_audit_capacity=normal_capacity,
                    safe_audit_capacity=safe_capacity,
                )


if __name__ == "__main__":
    unittest.main()
